src/ui/src/core_components/base/BaseControlBar.vue

```diff
@@ -0,0 +1,110 @@
+<template>
+	<div class="BaseControlBar">
+		<button
+			v-if="props.copyStructuredContent"
+			class="control-button"
+			@click="copyToClipboard({ text: props.copyStructuredContent })"
+		>
+			Copy JSON
+		</button>
+		<button
+			v-if="props.copyRawContent"
+			class="control-button"
+			@click="copyToClipboard({ text: props.copyRawContent })"
+		>
+			Copy
+		</button>
+	</div>
+</template>
+
+<script setup lang="ts">
+const props = defineProps<{
+	copyRawContent?: string;
+	copyStructuredContent?: string;
+}>();
+
+function setClipboardData<T = unknown>(
+	source: T & { clipboardData: DataTransfer | null | undefined },
+	{ text, html }: { text?: string; html?: string },
+): void {
+	if (text) {
+		source.clipboardData?.setData("text/plain", text);
+		source.clipboardData?.setData("Text", text); // IE mimetype
+	}
+
+	if (html) {
+		source.clipboardData?.setData("text/html", html);
+	}
+}
+
+function copyToClipboard({
+	text = "",
+	html = "",
+}: {
+	text?: string;
+	html?: string;
+}): boolean {
+	if (
+		(window as any)?.clipboardData &&
+		(window as any)?.clipboardData.setData
+	) {
+		// Internet Explorer-specific code path to prevent textarea being shown while dialog is visible.
+		setClipboardData<any>(window, { text, html });
+
+		return true;
+	} else if (
+		document.queryCommandSupported &&
+		document.queryCommandSupported("copy")
+	) {
+		const copyListener = (event: ClipboardEvent) => {
+			event.preventDefault();
+			setClipboardData<ClipboardEvent>(event, { text, html });
+		};
+
+		document.addEventListener("copy", copyListener, false);
+
+		const textarea = document.createElement("textarea");
+		textarea.textContent = text || html;
+		textarea.style.position = "fixed";
+		document.body.appendChild(textarea);
+		textarea.select();
+
+		try {
+			return document.execCommand("copy"); // Security exception may be thrown by some browsers.
+		} catch (ex) {
+			return false;
+		} finally {
+			document.body.removeChild(textarea);
+			document.removeEventListener("copy", copyListener, false);
+		}
```

Comment:
`document.execCommand` is deprecated API ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand)). You should use [Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API) like this

```ts
await navigator.clipboard.writeText(text);
```

You can even use this solution [with this polyfill](https://github.com/lgarron/clipboard-polyfill) to support old browsers

Reply:
Thanks for pointing this out!

It's [not deprecated](https://stackoverflow.com/a/78427138) in a traditional sense. The drama around it is long and could take 5 at least more years. I went with it since I don't see it being deprecated anytime soon and it just works for any browser. It just works for any content format, so the component is easily extensible.

However, it's a nice discussion point. I couldn't find a doc on what browsers do we actually support. If we are not hitting limitations with these [constraints](https://github.com/lgarron/clipboard-polyfill?tab=readme-ov-file#compatibility), I have no objections on using `navigator.clipboard`.

cc @ramedina86 @FabienArcellier